lxd/network_zones.go

```diff
@@ -196,50 +208,78 @@ func networkZoneAccessHandler(entitlement auth.Entitlement) func(d *Daemon, r *h
 func networkZonesGet(d *Daemon, r *http.Request) response.Response {
 	s := d.State()
 
-	requestProjectName := request.ProjectParam(r)
-	effectiveProjectName, _, err := project.NetworkZoneProject(s.DB.Cluster, requestProjectName)
-	if err != nil {
-		return response.SmartError(err)
+	allProjects := shared.IsTrue(request.QueryParam(r, "all-projects"))
+	requestProjectName := request.QueryParam(r, "project")
+
+	if allProjects && request.QueryParam(r, "project") != "" {
+		return response.BadRequest(errors.New("Cannot specify a project when requesting all projects"))
 	}
 
-	recursion := util.IsRecursionRequest(r)
+	var effectiveProjectName string
+	var err error
+	if !allProjects {
+		if requestProjectName == "" {
+			requestProjectName = api.ProjectDefaultName
+		}
+
+		effectiveProjectName, _, err = project.NetworkZoneProject(s.DB.Cluster, requestProjectName)
+		if err != nil {
+			return response.SmartError(err)
+		}
+
+		// If the request is project specific, then set effective project name in the request context so that the authorizer can generate the correct URL.
+		request.SetCtxValue(r, request.CtxEffectiveProjectName, effectiveProjectName)
+	}
 
-	var zoneNames []string
+	recursion := util.IsRecursionRequest(r)
 
+	var zoneNamesMap map[string]string
 	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
-		// Get list of Network zones.
-		zoneNames, err = tx.GetNetworkZonesByProject(ctx, effectiveProjectName)
+		if allProjects {
+			zoneNamesMap, err = tx.GetNetworkZones(ctx)
+		} else {
+			// Get list of Network zones.
+			zoneNames, err := tx.GetNetworkZonesByProject(ctx, effectiveProjectName)
+			if err != nil {
+				return err
+			}
+
+			zoneNamesMap = make(map[string]string, len(zoneNames))
+			for _, zoneName := range zoneNames {
+				zoneNamesMap[zoneName] = effectiveProjectName
+			}
+		}
 
 		return err
 	})
 	if err != nil {
 		return response.InternalError(err)
 	}
 
-	request.SetCtxValue(r, request.CtxEffectiveProjectName, effectiveProjectName)
 	userHasPermission, err := s.Authorizer.GetPermissionChecker(r.Context(), auth.EntitlementCanView, entity.TypeNetworkZone)
 	if err != nil {
 		return response.InternalError(err)
 	}
 
 	resultString := []string{}
 	resultMap := []api.NetworkZone{}
-	for _, zoneName := range zoneNames {
+	for zoneName, projectName := range zoneNamesMap {
 		if !userHasPermission(entity.NetworkZoneURL(requestProjectName, zoneName)) {
```

Comment:
This is still not making sense. The `requestProjectName` is only valid for project specific requests.

Reply:
I think this needs to be `projectName`. If the tests fail with this change, let's look in more detail at what the tests are doing and make sure they are correct.

Reply:
It appears that the TLS restrictions test suite fails with this change - I’ll look into this further.

Reply:
Still failing, here's where the failure occurs:

https://github.com/canonical/lxd/blob/da43979dab584872e6df2362834deb7fbd4187c3/test/suites/tls_restrictions.sh#L185-L187

I don't think it makes sense for the network zone to be visible in the project `blah` if it was created in the `default` project in this case since we are testing access for a restricted user.

Reply:
It does make sense in this case. The project `blah` does not have `features.networks.zones` enabled, so it uses the network zones from the default project.

Reply:
It does make sense in this case. The project `blah` does not have `features.networks.zones` enabled, so it uses the network zones from the default project.

Reply:
It does make sense in this case. The project `blah` does not have `features.networks.zones` enabled, so it uses the network zones from the default project.

Reply:
It looks like the initial value for `features.networks.zones` is set to `false`.

Reply:
Looks like we commented at the same time :)